src/DataLayouts/DataLayouts.jl

```diff
@@ -633,16 +633,19 @@ end
 
 function VF{S}(array::AbstractArray{T, 2}) where {S, T}
     check_basetype(T, S)
+    @assert size(array, 2) == typesize(T, S)
     VF{S, typeof(array)}(array)
 end
 
 function VF{S}(array::AbstractVector{T}) where {S, T}
+    check_basetype(T, S)
     @assert typesize(T, S) == 1
     VF{S}(reshape(array, (:, 1)))
 end
 
 function VF{S}(ArrayType, nelements) where {S}
```

Comment:
Might be better to use `::Type{A}` instead of `ArrayType` to ensure that the compiler specializes on `A`.

Reply:
You mean `function VF{S}(::Type{A}, nelements) where {S, A}`?

Reply:
it should still specialize in either case